### PR TITLE
Docker user directives cont

### PR DESCRIPTION
**Pull request by noah-paige**
### Feature or Bugfix

- Bugfix

### Detail
- Continuation of https://github.com/data-dot-all/dataall/pull/895/files

### Relates
- https://github.com/data-dot-all/dataall/pull/895/files

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by noah-paige** [comment]
**Fixes / Status Update:**
- Local Postgres DB 
    - Error Message
```
2024-01-11 16:22:50 The files belonging to this database system will be owned by user "cuser".
2024-01-11 16:22:50 This user must also own the server process.
2024-01-11 16:22:07 initdb: could not access directory "/var/lib/postgresql/data": Permission denied
```

Resolved By
        - Update to 13.12
        - Remove User Commands
        - Delete Container and Image and re build to get working with 13.12
	
- Local ElasticSearch 
    - Error Message
```
java.lang.IllegalStateException: failed to obtain node locks, tried [[/usr/share/elasticsearch/data]] with lock id [0]; maybe these locations are not writable or multiple nodes were started without increasing [node.max_local_storage_nodes] (was [1])?
```
Resolved By
        - Remove User Commands and ReBuild

- Docker Compose AWS Credentials
    - Have to change Mount Volumes for Docker User to get AWS Credentials locally
    - Otherwise getting Credentials Can Not be Located error from botocore because user can not access /.aws/credentials file
    - Resolved By
        - Update docker compose volume mounts

- VPC Facing AWS Deployment
    - Error building Image using frontend/docker/prod/Dockerfile

**Testing:**
Local data.all
- [x] Local docker compose up re-building all images
- [x] Access UI 
- [x] Create Environment
- [x] Create CDK Pipeline

_TODO: Fix VPC Facing AWS Deployment Issues
TODO: Test the above fixes with AWS Deployment_

File: docker-compose.yaml [hunk 1]
```diff
@@ -45,7 +46,7 @@ services:
       email_sender_id: "noreply@someawsdomain"
     volumes:
       - ./backend/dataall:/dataall
-      - $HOME/.aws/credentials:/root/.aws/credentials:ro
+      - $HOME/.aws/credentials:/home/cuser/.aws/credentials:ro
```
[the remark this answers is not in the document]
Comment (dlpzx) [reply docker-compose.yaml:49]: I run into the same issue

[the remark this answers is not in the document]
Comment (dlpzx) [reply docker-compose.yaml:49]: With this, environment creation locally works as usual

[the remark this answers is not in the document]
Comment (petrkalos) [reply docker-compose.yaml:49]: I am not sure why it works on Noah's setup but it really shouldn't, I think the most portable solution here is to make the UID of the newly created `cuser` equal to the UID of the host user.

I tested this idea and seems to be working fine on my setup, here is a diff that does the following

* The Dockerfile is creating a user with the id which is set in the ARG `CONTAINER_UID`
* `docker-compose` is passing the `CONTAINER_UID` variable from host's environment UID
* Users must run `export UID` before docker-compose

```diff
diff --git a/backend/docker/dev/Dockerfile b/backend/docker/dev/Dockerfile
index 51dccc05..e0c8881f 100644
--- a/backend/docker/dev/Dockerfile
+++ b/backend/docker/dev/Dockerfile
@@ -5,6 +5,7 @@ ARG NVM_VERSION=v0.37.2
 ARG PYTHON_VERSION=python3.9
 ARG CONTAINER_USER=cuser
 ARG CONTAINER_USER_GROUP=cuser
+ARG CONTAINER_UID
 
 
 
@@ -20,7 +21,7 @@ RUN dnf -y install -y \
 RUN dnf install $PYTHON_VERSION
 RUN dnf -y install python3-pip python3-devel git
 
-RUN groupadd -r ${CONTAINER_USER_GROUP} && useradd -m -r -g ${CONTAINER_USER_GROUP} ${CONTAINER_USER}
+RUN groupadd -r ${CONTAINER_USER_GROUP} && useradd -m -r -u ${CONTAINER_UID} -g ${CONTAINER_USER_GROUP} ${CONTAINER_USER}
 RUN useradd -m app
 RUN chown -R ${CONTAINER_USER}:root /usr/local
 RUN chown -R ${CONTAINER_USER}:root /tmp
diff --git a/docker-compose.yaml b/docker-compose.yaml
index 86223e57..6b10cfea 100644
--- a/docker-compose.yaml
+++ b/docker-compose.yaml
@@ -5,6 +5,8 @@ services:
     build:
       context: ./backend
       dockerfile: docker/dev/Dockerfile
+      args:
+        CONTAINER_UID: ${UID}
     entrypoint: /bin/bash -c 'aws configure set region "eu-west-1" &&. ~/.nvm/nvm.sh && uvicorn cdkproxymain:app --host 0.0.0.0 --port 2805 --reload'
     expose:
       - 2805
@@ -33,6 +35,8 @@ services:
     build:
       context: ./backend
       dockerfile: docker/dev/Dockerfile
+      args:
+        CONTAINER_UID: ${UID}
     entrypoint: /bin/bash -c "../build/wait-for-it.sh elasticsearch:9200 -t 30 && python3.9 local_graphql_server.py"
     expose:
       - 5000
```

[the remark this answers is not in the document]
Comment (noah-paige) [reply docker-compose.yaml:53]: @petrkalos @dlpzx - I added the above changes proposed by Petros to take in UID and create user with the same id 

Tested by running `export UID=$(id -u) && docker compose up --build -d` and linking a new environment successfully

Let me know if the latest changes also work for you all and if good to merge. Then can update documentation as well to change the local deploy guide to use the above docker compose command

[the remark this answers is not in the document]
Comment (petrkalos) [reply docker-compose.yaml:53]: Nice, with this change users will get a message `The "UID" variable is not set. Defaulting to a blank string.` if they don't set it and later on the build will fail.

FYI running `export UID` is enough as the variable already exists as a shell variable and have to be exported in the environment.

**Review by petrkalos — APPROVED** [review]

**Pull request merged**